src/ir/iteration.h

```diff
@@ -55,59 +56,88 @@ template<template<class, class> class Scanner> class AbstractChildIterator {
 
     void operator++() { index++; }
 
-    Expression* operator*() { return parent.children[index]; }
+    Expression*& operator*() {
+      assert(index < parent.children.size());
+
+      // The vector of children is in reverse order, as that is how
+      // wasm-delegations-fields works. To get the order of execution, reverse
+      // things.
+      return *parent.children[parent.children.size() - 1 - index];
+    }
   };
 
 public:
-  SmallVector<Expression*, 4> children;
+  // The vector of children in the order emitted by wasm-delegations-fields
+  // (which is in reverse execution order).
+  SmallVector<Expression**, 4> children;
 
   AbstractChildIterator(Expression* parent) {
-    struct Traverser : public PostWalker<Traverser> {
-      Expression* parent;
-      SmallVector<Expression*, 4>* children;
-
-      // We need to scan subchildren exactly once - just the parent.
-      bool scanned = false;
-
-      static void scan(Traverser* self, Expression** currp) {
-        if (!self->scanned) {
-          self->scanned = true;
-          Scanner<Traverser, UnifiedExpressionVisitor<Traverser>>::scan(self,
-                                                                        currp);
-        } else {
-          // This is one of the children. Do not scan further, just note it.
-          self->children->push_back(*currp);
-        }
-      }
-    } traverser;
-    traverser.parent = parent;
-    traverser.children = &children;
-    traverser.walk(parent);
+    auto* self = (Specific*)this;
+
+#define DELEGATE_ID parent->_id
+
+#define DELEGATE_START(id)                                                     \
+  auto* cast = parent->cast<id>();                                             \
+  WASM_UNUSED(cast);
+
+#define DELEGATE_GET_FIELD(id, name) cast->name
+
+#define DELEGATE_FIELD_CHILD(id, name) self->addChild(parent, &cast->name);
+
+#define DELEGATE_FIELD_OPTIONAL_CHILD(id, name)                                \
+  if (cast->name) {                                                            \
+    self->addChild(parent, &cast->name);                                       \
+  }
+
+#define DELEGATE_FIELD_INT(id, name)
+#define DELEGATE_FIELD_INT_ARRAY(id, name)
+#define DELEGATE_FIELD_LITERAL(id, name)
+#define DELEGATE_FIELD_NAME(id, name)
+#define DELEGATE_FIELD_NAME_VECTOR(id, name)
+#define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name)
+#define DELEGATE_FIELD_SCOPE_NAME_USE(id, name)
+#define DELEGATE_FIELD_SCOPE_NAME_USE_VECTOR(id, name)
+#define DELEGATE_FIELD_SIGNATURE(id, name)
+#define DELEGATE_FIELD_TYPE(id, name)
+#define DELEGATE_FIELD_ADDRESS(id, name)
+
+#include "wasm-delegations-fields.h"
```

Comment:
I think it would be good to reverse the children vector at the end because most uses of this end up iterating on `children` directly, iirc. In fact, perhaps we could get rid of `operator*` and `operator++` entirely, since it's not clear to me how those would effectively be used. What would the end condition for the iteration be?

Reply:
Someone might get the iterator for the end, and start iterating at the start, doing `++` until they reach the end? But I do think we do not have such users atm in the codebase.

Reversing the vector seems slightly risky to me - someone might not iterate through them all (even when iterating in a loop, they can break) so we'd be doing a linear pass (of memory stores, not free) unnecessarily. The subtraction seems cheaper. (Reading the size is not free though.)

Reply:
Aha, I had forgotten that in fact this is normally used like `for (auto it : ChildIterator(...)) `. LGTM!